src/java.desktop/unix/native/libawt_xawt/awt/screencast_pipewire.c

```diff
@@ -0,0 +1,806 @@
+    LOAD_SYMBOL(fp_pw_thread_loop_accept, "pw_thread_loop_accept");
+    LOAD_SYMBOL(fp_pw_thread_loop_start, "pw_thread_loop_start");
+    LOAD_SYMBOL(fp_pw_thread_loop_stop, "pw_thread_loop_stop");
+    LOAD_SYMBOL(fp_pw_thread_loop_destroy, "pw_thread_loop_destroy");
+    LOAD_SYMBOL(fp_pw_thread_loop_lock, "pw_thread_loop_lock");
+    LOAD_SYMBOL(fp_pw_thread_loop_unlock, "pw_thread_loop_unlock");
+    LOAD_SYMBOL(fp_pw_properties_new, "pw_properties_new");
+
+    return TRUE;
+
+    fail:
+        dlclose(pipewire_libhandle);
+        pipewire_libhandle = NULL;
+    return FALSE;
+}
+
+void storeRestoreToken(const gchar* token) {
+    JNIEnv* env = (JNIEnv *) JNU_GetEnv(jvm, JNI_VERSION_1_2);
+    DEBUG_SCREENCAST("saving token %s\n", token);
+    if (env) {
+        jstring jtoken = (*env)->NewStringUTF(env, token);
+        if ((*env)->ExceptionCheck(env)) {
+            (*env)->ExceptionClear(env);
+        }
+        if (!jtoken) {
+            return;
+        }
+        (*env)->CallStaticVoidMethod(env, screencastHelperClass,
+                                     storeTokenMethodID, jtoken);
+        if ((*env)->ExceptionCheck(env)) {
+            (*env)->ExceptionClear(env);
+        }
+
+        (*env)->DeleteLocalRef(env, jtoken);
+    } else {
+        DEBUG_SCREENCAST("⚠ Could not get env\n", NULL);
+    }
+}
+
+/*
+ * Class:     sun_awt_UNIXToolkit
+ * Method:    load_gtk
+ * Signature: (IZ)Z
+ */
+JNIEXPORT jboolean JNICALL Java_sun_awt_screencast_ScreencastHelper_loadPipewire(
+        JNIEnv *env, jclass cls, jboolean screencastDebug
+) {
+    DEBUG_SCREENCAST_ENABLED = screencastDebug;
+
+    if (!loadSymbols()) {
+        return FALSE;
+    }
+    screencastHelperClass = (*env)->NewGlobalRef(env, cls);
+    if (screencastHelperClass) {
+        storeTokenMethodID = (*env)->GetStaticMethodID(
+                env,
+                screencastHelperClass,
+                "storeToken",
+                "(Ljava/lang/String;)V"
+                );
+    } else {
+        DEBUG_SCREENCAST("⚠ @@@ screencastHelperClass %p\n",
+                         screencastHelperClass);
+    }
+
+    gboolean usable = initXdgDesktopPortal();
+    portalScreenCastCleanup();
+    return usable;
+}
+
+static void releaseToken(JNIEnv *env, jstring jtoken, const gchar *token) {
+    if (token) {
+        (*env)->ReleaseStringUTFChars(env, jtoken, token);
+    }
+}
+
+/*
+ * Class:     sun_awt_screencast_ScreencastHelper
+ * Method:    getRGBPixelsImpl
+ * Signature: (IIII[ILjava/lang/String;)I
+ */
+JNIEXPORT jint JNICALL Java_sun_awt_screencast_ScreencastHelper_getRGBPixelsImpl(
+        JNIEnv *env,
+        jclass cls,
+        jint jx,
+        jint jy,
+        jint jwidth,
+        jint jheight,
+        jintArray pixelArray,
+        jstring jtoken
+) {
+
+    GdkRectangle requestedArea = { jx, jy, jwidth, jheight};
+
+    const gchar *token = jtoken
+                         ? token = (*env)->GetStringUTFChars(env, jtoken, NULL)
```

Comment:
The variable `token` is unnecessarily used in its own initializer here.

Reply:
Fixed.